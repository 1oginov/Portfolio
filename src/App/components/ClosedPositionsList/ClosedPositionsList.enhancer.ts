import { compose, withProps } from 'recompose';

import * as C from 'Constants';
import { Position } from 'Portfolio/lib';
import { sortCollection } from 'Shared/lib';

import { Props } from './ClosedPositionsList';
import withSorter, { Props as WithSorterProps } from '../../enhancers/withSorter';
import withStockPositions, { Props as WithStockPositionsProps } from '../../enhancers/withStockPositions';
import { StockPosition } from '../../lib';

interface EnhancedProps {
  onPositionClick?: (positionId: string) => void;
  positions: Position[];
}

export default compose<Props, EnhancedProps>(
  withProps<EnhancedProps, EnhancedProps>(({ positions }) => ({
    positions: positions.filter(position => position.closeDate !== null),
  })),
  withStockPositions,
  withSorter(
    C.CLOSED_POSITIONS_LIST_SORTER_NAME,
    C.CLOSED_POSITIONS_LIST_SORTER_INITIAL_KEY,
    C.CLOSED_POSITIONS_LIST_SORTER_INITIAL_ORDER,
  ),
  withProps<Partial<Props>, EnhancedProps & WithStockPositionsProps & WithSorterProps>(({
    stockPositions, sorterKey, sorterOrder,
  }) => {
    let totalOpenSum = 0;
    let totalCloseSum = 0;

    stockPositions.forEach(position => {
      totalOpenSum += position.openSum;

      if (position.closeSum !== null) {
        totalCloseSum += position.closeSum;
      }
    });

    const totalClosePL = totalCloseSum - totalOpenSum;
    const totalClosePLPercent = totalClosePL / totalOpenSum;

    return {
      stockPositions: sortCollection(stockPositions, sorterKey as keyof StockPosition, sorterOrder),
      totalClosePL,
      totalClosePLPercent,
      totalCloseSum,
    };
  }),
);
