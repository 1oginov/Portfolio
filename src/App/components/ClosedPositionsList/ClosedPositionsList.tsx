import {
  List, ListItem, ListItemIcon, ListItemSecondaryAction, ListItemText,
} from '@material-ui/core';
import * as React from 'react';

import * as C from 'Constants';
import PositionDate from 'Portfolio/components/PositionDate';
import Money from 'Shared/components/Money';
import Percent from 'Shared/components/Percent';
import Sorter from 'Shared/components/Sorter';
import StockLogo from 'Stocks/components/StockLogo';

import { StockPosition } from '../../lib';

// TODO: Tests.

export interface Props {
  classes: { [key: string]: string };
  handleSorterKeyChange: (key: string) => void;
  handleSorterOrderChange: (order: 'asc' | 'desc') => void;
  onPositionClick?: (positionId: string) => void;
  sorterKey: string;
  sorterOrder: 'asc' | 'desc';
  stockPositions: StockPosition[];
  totalClosePL: number;
  totalClosePLPercent: number;
  totalCloseSum: number;
}

const ClosedPositionsList: React.FunctionComponent<Props> = ({
  classes, handleSorterKeyChange, handleSorterOrderChange, onPositionClick, stockPositions, sorterKey, sorterOrder,
  totalClosePL, totalClosePLPercent, totalCloseSum,
}: Props) => (
  <React.Fragment>

    <Money value={totalCloseSum} />
    <Money pl value={totalClosePL} />
    <Percent pl value={totalClosePLPercent} />

    <Sorter
      keys={C.CLOSED_POSITIONS_LIST_SORTER_KEYS}
      onKeyChange={handleSorterKeyChange}
      onOrderChange={handleSorterOrderChange}
      sorterKey={sorterKey}
      sorterOrder={sorterOrder}
    />

    <List className={classes.list} dense disablePadding>
      {stockPositions.map(({
        amount, closeDate, closePL, closePLAnnualPercent, closePLPercent, companyName, id, openDate, symbol,
      }) => (
        <ListItem button key={id} onClick={() => onPositionClick && onPositionClick(id)}>

          <ListItemIcon><StockLogo symbol={symbol} /></ListItemIcon>

          <ListItemText
            primary={companyName === null ? symbol : companyName}
            secondary={(
              <React.Fragment>
                {amount}
                {' @ '}
                {sorterKey === 'openDate' || closeDate === null
                  ? <PositionDate highlighted={sorterKey === 'openDate'} date={openDate} />
                  : <PositionDate highlighted={sorterKey === 'closeDate'} date={closeDate} />}
              </React.Fragment>
            )}
          />

          {sorterKey === 'closePLAnnualPercent'
            ? (
              <ListItemSecondaryAction className={classes.listItemSecondaryAction}>
                {closePLAnnualPercent !== null && <Percent pl value={closePLAnnualPercent} />}
              </ListItemSecondaryAction>
            ) : (
              <ListItemSecondaryAction className={classes.listItemSecondaryAction}>
                {closePL !== null && <Money highlighted={sorterKey === 'closePL'} pl value={closePL} />}
                <br />
                {closePLPercent !== null && (
                  <Percent highlighted={sorterKey === 'closePLPercent'} pl value={closePLPercent} />
                )}
              </ListItemSecondaryAction>
            )}

        </ListItem>
      ))}
    </List>

  </React.Fragment>
);

export default ClosedPositionsList;
