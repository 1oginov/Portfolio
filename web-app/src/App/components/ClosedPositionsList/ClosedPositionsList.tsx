import {
  List, ListItem, ListItemIcon, ListItemSecondaryAction, ListItemText,
} from '@material-ui/core';
import * as React from 'react';

import * as C from 'Constants';
import PositionDate from 'Portfolio/components/PositionDate';
import Money from 'Shared/components/Money';
import Percent from 'Shared/components/Percent';
import Sorter from 'Shared/components/Sorter';
import StockLogo from 'Stocks/components/StockLogo';

import { StockPosition } from '../../lib';

export interface Props {
  classes: { [key: string]: string };
  handleSorterKeyChange: (key: string) => void;
  handleSorterOrderChange: (order: 'asc' | 'desc') => void;
  onPositionClick?: (positionId: string) => void;
  sorterKey: string;
  sorterOrder: 'asc' | 'desc';
  stockPositions: StockPosition[];
}

const ClosedPositionsList: React.FunctionComponent<Props> = ({
  classes, handleSorterKeyChange, handleSorterOrderChange, onPositionClick, stockPositions, sorterKey, sorterOrder,
}: Props) => (
  <>

    <Sorter
      className={classes.sorter}
      keys={C.CLOSED_POSITIONS_LIST_SORTER_KEYS}
      onKeyChange={handleSorterKeyChange}
      onOrderChange={handleSorterOrderChange}
      sorterKey={sorterKey}
      sorterOrder={sorterOrder}
    />

    <List className={classes.list} dense disablePadding>
      {stockPositions.map(({
        amount, closeDate, closePL, closePLAnnualPercent, closePLPercent, companyName, id, openDate, symbol,
      }) => (
        <ListItem button key={id} onClick={() => onPositionClick && onPositionClick(id)}>

          <ListItemIcon><StockLogo symbol={symbol} /></ListItemIcon>

          <ListItemText
            primary={companyName === null ? symbol : companyName}
            secondary={(
              <>
                {amount}
                {' '}
                @
                {' '}
                {sorterKey === 'openDate' || closeDate === null
                  ? <PositionDate highlighted={sorterKey === 'openDate'} date={openDate} />
                  : <PositionDate highlighted={sorterKey === 'closeDate'} date={closeDate} />}
              </>
            )}
          />

          {sorterKey === 'closePLAnnualPercent'
            ? (
              <ListItemSecondaryAction className={classes.listItemSecondaryAction}>
                {closePLAnnualPercent !== null && <Percent pl value={closePLAnnualPercent} />}
              </ListItemSecondaryAction>
            ) : (
              <ListItemSecondaryAction className={classes.listItemSecondaryAction}>
                {closePL !== null && <Money highlighted={sorterKey === 'closePL'} pl value={closePL} />}
                <br />
                {closePLPercent !== null && (
                  <Percent highlighted={sorterKey === 'closePLPercent'} pl value={closePLPercent} />
                )}
              </ListItemSecondaryAction>
            )}

        </ListItem>
      ))}
    </List>

  </>
);

export default ClosedPositionsList;
