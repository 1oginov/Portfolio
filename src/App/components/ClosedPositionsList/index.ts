import ClosedPositionsList from './ClosedPositionsList';
import enhance from './ClosedPositionsList.enhancer';
import style from './ClosedPositionsList.styles';

export default style(enhance(ClosedPositionsList));
