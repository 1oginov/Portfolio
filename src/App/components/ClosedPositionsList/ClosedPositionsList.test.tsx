import { shallow } from 'enzyme';
import classes from 'identity-obj-proxy';
import * as React from 'react';

import ClosedPositionsList from './ClosedPositionsList';

it('matches snapshot', () => {
  const wrapper = shallow(
    <ClosedPositionsList
      classes={classes}
      handleSorterKeyChange={() => undefined}
      handleSorterOrderChange={() => undefined}
      sorterKey="openPrice"
      sorterOrder="asc"
      stockPositions={[]}
      totalClosePL={100}
      totalClosePLPercent={1}
      totalCloseSum={1000}
    />,
  );

  expect(wrapper).toMatchSnapshot();
});
