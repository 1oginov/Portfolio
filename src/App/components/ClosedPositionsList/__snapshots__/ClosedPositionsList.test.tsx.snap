// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<Fragment>
  <WithStyles(Money)
    value={1000}
  />
  <WithStyles(Money)
    pl={true}
    value={100}
  />
  <WithStyles(Percent)
    pl={true}
    value={1}
  />
  <withState(withProps(Sorter))
    keys={
      Array [
        Object {
          "key": "companyName",
          "label": "Company Name",
        },
        Object {
          "key": "openDate",
          "label": "Open Date",
        },
        Object {
          "key": "closeDate",
          "label": "Close Date",
        },
        Object {
          "key": "closePL",
          "label": "PL",
        },
        Object {
          "key": "closePLPercent",
          "label": "PL%",
        },
        Object {
          "key": "closePLAnnualPercent",
          "label": "Annual PL%",
        },
      ]
    }
    onKeyChange={[Function]}
    onOrderChange={[Function]}
    sorterKey="openPrice"
    sorterOrder="asc"
  />
  <WithStyles(ForwardRef(List))
    className="list"
    dense={true}
  />
</Fragment>
`;
